pkg/distributor/ratestore.go

```diff
@@ -1,25 +1,234 @@
+	fs.DurationVar(&cfg.IngesterReqTimeout, prefix+".ingester-request-timeout", time.Second, "Timeout for communication between distributors and ingesters when updating rates")
+}
+
+type ingesterClient struct {
+	addr   string
+	client logproto.StreamDataClient
+}
+
+type overrides interface {
+	AllByUserID() map[string]*validation.Limits
+}
+
+type rateStore struct {
+	services.Service
+
+	ring                   ring.ReadRing
+	clientPool             poolClientFactory
+	rates                  map[uint64]int64
+	rateLock               sync.RWMutex
+	rateCollectionInterval time.Duration
+	ingesterTimeout        time.Duration
+	maxParallelism         int
+	rateRefreshFailures    *prometheus.CounterVec
+	refreshDuration        *instrument.HistogramCollector
+	overrides              overrides
+}
+
+func NewRateStore(cfg RateStoreConfig, r ring.ReadRing, cf poolClientFactory, o overrides, registerer prometheus.Registerer) *rateStore { //nolint
+	s := &rateStore{
+		ring:                   r,
+		clientPool:             cf,
+		rateCollectionInterval: cfg.StreamRateUpdateInterval,
+		maxParallelism:         cfg.MaxParallelism,
+		ingesterTimeout:        cfg.IngesterReqTimeout,
+		overrides:              o,
+		rateRefreshFailures: promauto.With(registerer).NewCounterVec(prometheus.CounterOpts{
+			Namespace: "loki",
+			Name:      "rate_store_refresh_failures_total",
+			Help:      "The total number of failed attempts to refresh the distributor's view of stream rates",
+		}, []string{"source"}),
+		refreshDuration: instrument.NewHistogramCollector(
+			promauto.With(registerer).NewHistogramVec(
+				prometheus.HistogramOpts{
+					Namespace: "loki",
+					Name:      "rate_store_refresh_duration_seconds",
+					Help:      "Time spent refreshing the rate store",
+					Buckets:   prometheus.DefBuckets,
+				}, instrument.HistogramCollectorBuckets,
+			),
+		),
+	}
+
+	s.Service = services.
+		NewTimerService(s.rateCollectionInterval, s.instrumentedUpdateAllRates, s.instrumentedUpdateAllRates, nil).
+		WithName("rate store")
+
+	return s
+}
+
+func (s *rateStore) instrumentedUpdateAllRates(ctx context.Context) error {
+	if !s.anySharingEnabled() {
+		return nil
+	}
+
+	return instrument.CollectedRequest(ctx, "GetAllStreamRates", s.refreshDuration, instrument.ErrorCode, s.updateAllRates)
+}
+
+func (s *rateStore) updateAllRates(ctx context.Context) error {
+	clients, err := s.getClients()
+	if err != nil {
+		level.Error(util_log.Logger).Log("msg", "error getting ingester clients", "err", err)
+		s.rateRefreshFailures.WithLabelValues("ring").Inc()
+		return nil // Don't fail the service because we have an error getting the clients once
+	}
+
+	streamRates := s.getRates(ctx, clients)
+	rates := s.aggregateByShard(streamRates)
+
+	s.rateLock.Lock()
+	defer s.rateLock.Unlock()
+	s.rates = rates
+
+	return nil
+}
+
+func (s *rateStore) anySharingEnabled() bool {
```

Comment:
nit (sharing -> sharding)

in case it was unwanted make sure to also update it at line 101 (mentioning it as just applying this suggestion will result in broken code)
```suggestion
func (s *rateStore) anyShardingEnabled() bool {
```

Reply:
Thank you! 🤦‍♂️